### PR TITLE
remove eip6110flag for validator index cache

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

 Other

**What does this PR do? Why is it needed?**

because this cache is used for not only EIP6110 but also some others like 7521 we would like a way to use it out of the box without the need for a feature flag. longer term we need to revisit if we should solely depend on the cache or utilize the existing state features.

continuation of https://github.com/prysmaticlabs/prysm/pull/14146

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

### CODE REVIEW

File: beacon-chain/state/state-native/getters_validator.go [hunk 1]
```diff
@@ -179,7 +179,7 @@ func (b *BeaconState) ValidatorIndexByPubkey(key [fieldparams.BLSPubkeyLength]by
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
-	if features.Get().EIP6110ValidatorIndexCache {
+	if b.Version() >= version.Electra {
```
Comment (james-prysm) [inline beacon-chain/state/state-native/getters_validator.go:182]: what do we do about validatorMultiValue and the valMapHandler?

**Comment by prestonvanloon** [comment]
Do you need the feature flag for testing?

Consider renaming the flag if the justification is that other eip features use ValidatorByPubkey

**Comment by james-prysm** [comment]
> Do you need the feature flag for testing?
> 
> Consider renaming the flag if the justification is that other eip features use ValidatorByPubkey

I believe we actually need it now not behind a feature flag, asking the team for more thoughts and review

File: beacon-chain/state/state-native/state_trie.go [hunk 2]
```diff
@@ -754,11 +759,12 @@ func InitializeFromProtoUnsafeElectra(st *ethpb.BeaconStateElectra) (state.Beaco
...
-		dirtyIndices:     make(map[types.FieldIndex][]uint64, fieldCount),
-		stateFieldLeaves: make(map[types.FieldIndex]*fieldtrie.FieldTrie, fieldCount),
-		rebuildTrie:      make(map[types.FieldIndex]bool, fieldCount),
-		valMapHandler:    stateutil.NewValMapHandler(st.Validators),
+		dirtyFields:         make(map[types.FieldIndex]bool, fieldCount),
+		dirtyIndices:        make(map[types.FieldIndex][]uint64, fieldCount),
+		stateFieldLeaves:    make(map[types.FieldIndex]*fieldtrie.FieldTrie, fieldCount),
+		rebuildTrie:         make(map[types.FieldIndex]bool, fieldCount),
+		valMapHandler:       stateutil.NewValMapHandler(st.Validators),
+		validatorIndexCache: newFinalizedValidatorIndexCache(), //only populates when finalizing, otherwise it falls back to validator set
```
Comment (rkapka) [inline beacon-chain/state/state-native/state_trie.go:767]: Shouldn't you copy the comment to all other state types?

Comment (saolyn) [reply beacon-chain/state/state-native/state_trie.go:767]: Is this an electra specific behaviour?

Comment (james-prysm) [reply beacon-chain/state/state-native/state_trie.go:767]: it's electra specific but i was afraid there might be panics if not initialized

Comment (rkapka) [reply beacon-chain/state/state-native/state_trie.go]: it's not Electra-specific, this field is present in all states

File: beacon-chain/state/state-native/setters_misc.go [hunk 3]
```diff
@@ -110,7 +109,7 @@ func (b *BeaconState) SetHistoricalRoots(val [][]byte) error {
 
 // SaveValidatorIndices save validator indices of beacon chain to cache
 func (b *BeaconState) SaveValidatorIndices() {
-	if !features.Get().EIP6110ValidatorIndexCache {
+	if b != nil && b.Version() < version.Electra {
```
Comment (rkapka) [inline beacon-chain/state/state-native/setters_misc.go:112]: If `b` is nil, the function will continue and panic at the next line. It should be `b == nil ||` instead. Either way I think ni-checking receivers is too paranoid.

**Review by rkapka — APPROVED** [review]

**Pull request merged**